client/tx/tx.go

```diff
@@ -36,3 +53,311 @@ type (
+		for i, gp := range txf.gasPrices {
+			fee := gp.Amount.Mul(glDec)
+			fees[i] = sdk.NewCoin(gp.Denom, fee.Ceil().RoundInt())
+		}
+	}
+
+	tx := txf.txGenerator.NewTx()
+	tx.SetFee(auth.NewStdFee(txf.gas, fees))
+	tx.SetMsgs(msgs...)
+	tx.SetMemo(txf.memo)
+	tx.SetSignatures()
+
+	return tx, nil
+}
+
+// BuildSimTx creates an unsigned tx with an empty single signature and returns
+// the encoded transaction or an error if the unsigned transaction cannot be
+// built.
+func BuildSimTx(txf Factory, msgs ...sdk.Msg) ([]byte, error) {
+	tx, err := BuildUnsignedTx(txf, msgs...)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create an empty signature literal as the ante handler will populate with a
+	// sentinel pubkey.
+	tx.SetSignatures(auth.NewStdSignature(nil, nil))
+
+	return tx.Marshal()
+}
+
+// CalculateGas simulates the execution of a transaction and returns the
+// simulation response obtained by the query and the adjusted gas amount.
+func CalculateGas(
+	queryFunc func(string, []byte) ([]byte, int64, error), txf Factory, msgs ...sdk.Msg,
+) (sdk.SimulationResponse, uint64, error) {
+
+	txBytes, err := BuildSimTx(txf, msgs...)
+	if err != nil {
+		return sdk.SimulationResponse{}, 0, err
+	}
+
+	bz, _, err := queryFunc("/app/simulate", txBytes)
+	if err != nil {
+		return sdk.SimulationResponse{}, 0, err
+	}
+
+	var simRes sdk.SimulationResponse
+	if err := json.Unmarshal(bz, &simRes); err != nil {
+		return sdk.SimulationResponse{}, 0, err
+	}
+
+	return simRes, uint64(txf.GasAdjustment() * float64(simRes.GasUsed)), nil
+}
+
+// PrepareFactory ensures the account defined by ctx.GetFromAddress() exists and
+// if the account number and/or the account sequence number are zero (not set),
+// they will be queried for and set on the provided Factory. A new Factory with
+// the updated fields will be returned.
+func PrepareFactory(ctx context.CLIContext, txf Factory) (Factory, error) {
+	from := ctx.GetFromAddress()
+
+	if err := txf.accountRetriever.EnsureExists(from); err != nil {
+		return txf, err
+	}
+
+	initNum, initSeq := txf.accountNumber, txf.sequence
+	if initNum == 0 || initSeq == 0 {
+		num, seq, err := txf.accountRetriever.GetAccountNumberSequence(from)
+		if err != nil {
+			return txf, err
+		}
+
+		if initNum == 0 {
+			txf = txf.WithAccountNumber(num)
+		}
+		if initSeq == 0 {
+			txf = txf.WithSequence(seq)
+		}
+	}
+
+	return txf, nil
+}
+
+// Sign signs a given tx with the provided name and passphrase. If the Factory's
+// Keybase is not set, a new one will be created based on the client's backend.
+// The bytes signed over are canconical. The resulting signature will be set on
+// the transaction. Finally, the marshaled transaction is returned. An error is
+// returned upon failure.
+//
+// Note, It is assumed the Factory has the necessary fields set that are required
+// by the CanonicalSignBytes call.
+func Sign(txf Factory, name, passphrase string, tx ClientTx) ([]byte, error) {
+	if txf.keybase == nil {
```

Comment:
This is a bit of concern as this is the origin of problems with `gentx` (and other commands) that prompt users for password multiple times when using file-backed keyrings (e.g. https://github.com/cosmos/cosmos-sdk/issues/5771)
I'd amend this check as the function should assume that the factory is correctly initialised and can sign. Maybe we should just check whether keybase is nil, and if so return an error?

Reply:
Another bug: https://github.com/cosmos/cosmos-sdk/issues/5403

Reply:
I'd understand if you reckoned that this is out of this PR's scope. It does not necessarily have to be addressed here, so feel free to ignore this

Reply:
Sure, I'll just return an error instead.